service/src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java

```diff
@@ -630,4 +641,35 @@ private JobBuilder commonDeletionJobBuilder(
         .addParameter(ResourceKeys.RESOURCE_ID, resourceId.toString())
         .addParameter(JobMapKeys.RESULT_PATH.getKeyName(), resultPath);
   }
+
+  /**
+   * For async resource creation, we do not want to return to the caller until the resource row is
+   * in the database and, thus, visible to enumeration. This method waits for either the row to show
+   * up (the expected success case) or the job to complete (the expected error case). If one of
+   * those doesn't happen in the retry window, we throw SERVICE_UNAVAILABLE. The theory is for it
+   * not to complete, either WSM is so busy that it cannot schedule the flight or something bad has
+   * happened. Either way, SERVICE_UNAVAILABLE seems like a reasonable response.
+   *
+   * @param workspaceId workspace of the resource create
+   * @param resourceId id of resource being created
+   * @param jobId id of the create flight.
+   */
+  private void waitForResourceOrJob(UUID workspaceId, UUID resourceId, String jobId) {
+    try {
+      for (int i = 0; i < RESOURCE_ROW_MAX_RETRIES; i++) {
+        if (resourceDao.checkForResourceRow(workspaceId, resourceId)) {
+          return;
+        }
+        FlightState flightState = jobService.getStairway().getFlightState(jobId);
+        if (flightState.getCompleted().isPresent()) {
```

Comment:
Are you sure there's no race condition where we get here and the job completed successfully? Should we predicate the first branch on the  job being successfully started but not finished?

Reply:
It is possible that we get here and the job completed successfully. That is OK. We return the job status to the client. The client calls the `/result` endpoint and gets the successful result.
It is unexpected, because we would expect to see the initial step complete before the whole flight completes, but it could happen.
Does that make sense? Perhaps I can resolve this by adding more commentary above?

Reply:
That makes sense. It smelled like there might be coupling or an assumption about timing, but it sounds like you've worked that out. A blurb about this possibility in a comment couldn't hurt I suppose. Thanks.